[Feat] Prometheus metric export (#134)
ml-energy/zeus: docs/measure/index.md

```
5. Start Prometheus:
```sh
./prometheus --config.file=prometheus.yml
```
6. Visit http://localhost:9090 in your browser, or use curl http://localhost:9090/api/v1/targets
7. Verify Prometheus is running by visiting http://localhost:9090 in your browser.

Zeus allows you to monitor energy and power consumption through different metrics, such as Histograms, Counters, and Gauges, which can be pushed to a Prometheus Push Gateway for further analysis. 

---

[`EnergyHistogram`][zeus.metric.EnergyHistogram] records energy consumption data for GPUs, CPUs, and DRAM in Prometheus Histograms. This is ideal for observing how often energy usage falls within specific ranges.

You can customize the bucket ranges for each component (GPU, CPU, and DRAM), or let Zeus use default ranges.

```python hl_lines="2 5-15"

from zeus.metric import EnergyHistogram

if __name__ == "__main__":
    # Initialize EnergyHistogram
    energy_histogram = EnergyHistogram(
        cpu_indices=[0], 
        gpu_indices=[0], 
        prometheus_url='http://localhost:9091', 
        job='training_energy_histogram'
    )

    for epoch in range(100):
        # Start monitoring energy for the entire epoch

>         energy_histogram.begin_window(f"epoch_{epoch}_energy")

Reviewer: This got me confused. Why is it okay for the window name to be different on each epoch? I thought the window name should always be the same (e.g., `training_epoch`), and each time when a window with exactly that name is measured (in this example snippet, every epoch), it will add a data point to the histogram metric uniquely identified by that window name.

Reviewer: This is the largest concern of this review round. I see below that you also have:

```python
energy_histogram.begin_window(f"epoch_{epoch}_step_{step_idx}_energy")
```

And you're using the same instance of `energy_histogram`. I hope I'm wrong, but this has to mean that you're bunching together energy measurements from both training epoch and training step into the same `Histogram` metric with just different `window` labels, aren't you? If so, this defeats the whole purpose of histogram monitoring. The result of this example should be two completely independent `Histogram` metrics, one for monitoring training epochs (one datapoint added whenever `end_window("training_epoch")` is called) and another for monitoring training steps (one datapoint added whenever `end_window("training_step")` is called).

        avg_energy = sum(map(lambda m: m.total_energy, steps)) / len(steps)
        print(f"One step takes {avg_time} s and {avg_energy} J for the CPU.")
```

> ## Metric Monitoring

Reviewer: We need to have a section on our assumptions on Prometheus. For instance, we assume that the Prometheus Push Gateway is deployed, so we need to talk about that and have a brief section for how to do that or a link to an official documentation. Are there other assumptions?

Zeus allows you to monitor energy and power consumption through different metrics, such as Histograms, Counters, and Gauges, which can be pushed to a Prometheus Push Gateway for further analysis. 

[`EnergyHistogram`][zeus.metric.EnergyHistogram] records energy consumption data for GPUs, CPUs, and DRAM in Prometheus Histograms. This is useful for observing how frequently energy usage reaches certain levels.

You can customize the bucket ranges for each component (GPU, CPU, and DRAM), or let Zeus use default ranges.

> ```python hl_lines="2 5-7"

Reviewer: ```suggestion
```python hl_lines="2 5-15"
```
You can preview documentation with `bash scripts/preview_docs.sh`. The script will also report missing docstrings and stuff.

from zeus.monitor import ZeusMonitor
from zeus.metric import EnergyHistogram

if __name__ == "__main__":
    # Initialize EnergyHistogram with custom bucket ranges
    histogram_metric = EnergyHistogram(

>         energy_monitor=ZeusMonitor,

Reviewer: This should be an actual instance of a ZeusMonitor.

        prometheus_url='http://localhost:9091',
        job='energy_histogram_job',
        bucket_ranges={
            "gpu": [10.0, 25.0, 50.0, 100.0],
            "cpu": [5.0, 10.0, 25.0, 50.0],
            "dram": [1.0, 2.5, 5.0, 10.0]

>         }

Reviewer: I think we can remove this from the example for simplicity. We also have a separate `!!! tip` box for this argument.

    )

    histogram_metric.begin_window("histogram_test")
    # Perform tasks

>     histogram_metric.end_window("histogram_test")

Reviewer: Give it a more concrete example. I would write a simplified training loop in PyTorch (see above in this file for reference) and then have the histogram metric collect a histogram of the energy consumption of one training epoch.

```
## Metric Monitoring

Zeus allows for efficient monitoring of energy and power consumption for GPUs, CPUs, and DRAM using Prometheus. It tracks key metrics such as energy usage, power draw, and cumulative consumption. Users can define measurement windows to track energy usage for specific operations, enabling granular analysis and optimization.

!!! Assumption

>     A Prometheus Push Gateway must be deployed and accessible at the URL provided in your Zeus configuration. **This ensures that metrics collected by Zeus can be pushed to Prometheus.**

Reviewer: ```suggestion
    A [Prometheus Push Gateway](https://prometheus.io/docs/instrumenting/pushing/) must be deployed and accessible. This ensures that metrics collected by Zeus can be pushed to Prometheus.
```

6. Visit http://localhost:9090 in your browser, or use curl http://localhost:9090/api/v1/targets
7. Verify Prometheus is running by visiting http://localhost:9090 in your browser.

### Metric Name Construction

Zeus organizes metrics using **static metric names** and **dynamic labels** for flexibility and ease of querying in Prometheus. Metric names are static and cannot be overridden, but users can customize the context of the metrics by naming the window when using `begin_window()` and `end_window()`.

#### Metric Name
- For Histogram: `energy_monitor_{component}_energy_joules`
- For Counter: `energy_monitor_{component}_energy_joules`

> - For Gauge: `power_monitor_gpu_power_watts`

Reviewer: ```suggestion
- For Gauge: `power_monitor_{component}_power_watts`

Note that Gauge only supports the GPU component at the moment. Tracking issue: [#128](https://github.com/ml-energy/zeus/issues/128)

```

component: gpu, cpu, or dram

#### Labels
- window: The user-defined window name provided during `begin_window()` and `end_window()` (e.g., `energy_histogram.begin_window(f"epoch_energy")`).

> - index: The GPU index (e.g., `0` for GPU 0).

Reviewer: ```suggestion
- index: The index of the device (e.g., `0` for GPU 0).
```

    for epoch in range(100):
        # Start monitoring energy for the entire epoch
        energy_histogram.begin_window("epoch_energy")
        # Perform epoch-level operations 
        train_one_epoch(train_loader, model, optimizer, criterion, epoch, args)
        acc1 = validate(val_loader, model, criterion, args)
        # End monitoring energy for the epoch
        energy_histogram.end_window("epoch_energy")
        print(f"Epoch {epoch} completed. Validation Accuracy: {acc1}%")

```
You can use the `begin_window` and `end_window` methods to define a measurement window, similar to other ZeusMonitor operations. Energy consumption data will be recorded for the entire duration of the window.

!!! Tip 
    You can customize the bucket ranges for GPUs, CPUs, and DRAM during initialization to tailor the granularity of energy monitoring. For example:
    ```python hl_lines="2 5-15"
    energy_histogram = EnergyHistogram(
        cpu_indices=[0], 
        gpu_indices=[0], 

>         prometheus_url='http://localhost:9091', 
>         job='training_energy_histogram',

Reviewer: ```suggestion
        prometheus_url="http://localhost:9091", 
        job="training_energy_histogram",
```

>         gpu_bucket_range = [10.0, 25.0, 50.0, 100.0],
>         cpu_bucket_range = [5.0, 15.0, 30.0, 50.0],
>         dram_bucket_range = [2.0, 8.0, 20.0, 40.0],

Reviewer: ```suggestion
        gpu_bucket_range=[10.0, 25.0, 50.0, 100.0],
        cpu_bucket_range=[5.0, 15.0, 30.0, 50.0],
        dram_bucket_range=[2.0, 8.0, 20.0, 40.0],
```

    )
    ```

If no custom `bucket ranges` are specified, Zeus uses these default ranges:

> ```
> - GPU: [50.0, 100.0, 200.0, 500.0, 1000.0]
> - CPU: [10.0, 20.0, 50.0, 100.0, 200.0]
> - DRAM: [5.0, 10.0, 20.0, 50.0, 150.0]
> ```

Reviewer: ```suggestion

- GPU: `[50.0, 100.0, 200.0, 500.0, 1000.0]`
- CPU: `[10.0, 20.0, 50.0, 100.0, 200.0]`
- DRAM: `[5.0, 10.0, 20.0, 50.0, 150.0]`
```

```
The `update_period` parameter defines how often the power datas are updated and pushed to Prometheus.

### How to Query Metrics in Prometheus

> #### Query to View Energy for a Specific Window
> ```promql
> energy_monitor_gpu_energy_joules{window="epoch_energy"}
> ```
> #### Query to Sum Energy for a Specific Window
> ```promql
> sum(energy_monitor_gpu_energy_joules) by (window)
> ```
> #### Query to Sum Energy for Specific GPU Across All Windows
> ```promql
> sum(energy_monitor_gpu_energy_joules{index="0"})
> ```

Reviewer: ````suggestion
Energy for a specific window:
```promql
energy_monitor_gpu_energy_joules{window="epoch_energy"}
```

Sum of energy for a specific window:
```promql
sum(energy_monitor_gpu_energy_joules) by (window)
```

Sum of energy for specific GPU across all windows:
```promql
sum(energy_monitor_gpu_energy_joules{index="0"})
```
````
